test/serviceprofiles/serviceprofiles_test.go

```diff
@@ -0,0 +1,161 @@
+package serviceprofiles
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/linkerd/linkerd2/testutil"
+)
+
+var TestHelper *testutil.TestHelper
+
+type testCase struct {
+	args           []string
+	deployName     string
+	expectedRoutes []string
+	namespace      string
+	sourceName     string
+	spName         string
+}
+
+func TestMain(m *testing.M) {
+	TestHelper = testutil.NewTestHelper()
+	os.Exit(m.Run())
+}
+
+func TestServiceProfiles(t *testing.T) {
+
+	testNamespace := TestHelper.GetTestNamespace("serviceprofile-test")
+	out, stderr, err := TestHelper.LinkerdRun("inject", "testdata/tap_application.yaml")
+	if err != nil {
+		t.Fatalf("'linkerd %s' command failed with %s: %s\n", "inject", err.Error(), stderr)
+	}
+
+	out, err = TestHelper.KubectlApply(out, testNamespace)
+	if err != nil {
+		t.Fatalf("kubectl apply command failed\n%s", out)
+	}
+
+	// wait for deployments to start
+	for _, deploy := range []string{"t1", "t2", "t3", "gateway"} {
+		if err := TestHelper.CheckPods(testNamespace, deploy, 1); err != nil {
+			t.Error(err)
+		}
+
+		if err := TestHelper.CheckDeployment(testNamespace, deploy, 1); err != nil {
+			t.Error(fmt.Errorf("Error validating deployment [%s]:\n%s", deploy, err))
+		}
+	}
+
+	testCases := []testCase{
+		{
+			sourceName:     "tap",
+			namespace:      testNamespace,
+			deployName:     "deploy/t1",
+			spName:         "t1-svc",
+			expectedRoutes: []string{"POST /buoyantio.bb.TheService/theFunction", "[DEFAULT]"},
+		},
+		{
+			sourceName:     "open-api",
+			namespace:      testNamespace,
+			spName:         "t3-svc",
+			deployName:     "deploy/t3",
+			expectedRoutes: []string{"DELETE /testpath", "GET /testpath", "PATCH /testpath", "POST /testpath", "[DEFAULT]"},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc // pin
+		t.Run(tc.sourceName, func(t *testing.T) {
+			routes, err := getRoutes(tc.deployName, tc.namespace)
+			if err != nil {
+				t.Fatalf("routes command failed: %s\n", err)
+			}
+
+			initialExpectedRoutes := []string{"[DEFAULT]"}
+
+			assertExpectedRoutes(initialExpectedRoutes, routes, t)
+
+			sourceFlag := fmt.Sprintf("--%s", tc.sourceName)
+			cmd := []string{"profile", "--namespace", tc.namespace, tc.spName, sourceFlag}
+			if tc.sourceName == "tap" {
+				tc.args = []string{
+					tc.deployName,
+					"--tap-route-limit",
+					"5",
+					"--tap-duration",
+					"10s",
+				}
+			}
+
+			if tc.sourceName == "open-api" {
+				tc.args = []string{
+					"testdata/t3.swagger",
+				}
+			}
+
+			cmd = append(cmd, tc.args...)
+			out, stderr, err := TestHelper.LinkerdRun(cmd...)
+			if err != nil {
+				t.Fatalf("'linkerd %s' command failed with %s: %s\n", cmd, err.Error(), stderr)
```

Comment:
this failed on my local system with:
```bash
$ go test -v ./test/serviceprofiles -integration-tests -linkerd `pwd`/bin/linkerd
=== RUN   TestServiceProfiles
=== RUN   TestServiceProfiles/tap
=== RUN   TestServiceProfiles/open-api
--- FAIL: TestServiceProfiles (35.05s)
    --- FAIL: TestServiceProfiles/tap (10.43s)
        serviceprofiles_test.go:102: 'linkerd [profile --namespace l5d-integration-serviceprofile-test t1-svc --tap deploy/t1 --tap-route-limit 5 --tap-duration 10s]' command failed with exit status 1: Error: Tap duration exceeded, try increasing --tap-duration
            Usage:
```

i'm guessing it's a timing issue between slow-cooker's `sleep 15` at startup, and the `profile --tap --tap-duration 10s` not seeing any requests.

i think you can mitigate this by increasing `--tap-duration` to `25s`, but also decrease `--tap-route-limit` to `1`, so the command should complete as soon as it sees the first request.

```diff
diff --git a/test/serviceprofiles/serviceprofiles_test.go b/test/serviceprofiles/serviceprofiles_test.go
index 34803828..c3fadc26 100644
--- a/test/serviceprofiles/serviceprofiles_test.go
+++ b/test/serviceprofiles/serviceprofiles_test.go
@@ -84,9 +84,9 @@ func TestServiceProfiles(t *testing.T) {
                                tc.args = []string{
                                        tc.deployName,
                                        "--tap-route-limit",
-                                       "5",
+                                       "1",
                                        "--tap-duration",
-                                       "10s",
+                                       "25s",
                                }
                        }
```